runtime/doc/treesitter.txt

```diff
@@ -678,24 +708,12 @@ inspect_language({lang})          *vim.treesitter.language.inspect_language()*
     Return: ~
         (table)
 
-                                  *vim.treesitter.language.require_language()*
-require_language({lang}, {path}, {silent}, {symbol_name})
-    Asserts that a parser for the language {lang} is installed.
-
-    Parsers are searched in the `parser` runtime directory, or the provided
-    {path}
+register({lang}, {filetype})              *vim.treesitter.language.register()*
```

Comment:
Would an interface like `vim.treesitter.config()` similar to `vim.diagnostic.config()` make sense? Perhaps that helps discoverability.

Reply:
Maybe. We could have something like: `vim.treesitter.config([lang}, {opts})`:

```lua
vim.treesitter.config('bash', {
  filetype = {'bash', 'sh'},
  ???
})
```

I think this only works though if a specific language can be configured once. This won't be ideal If a user also wants to further setup the language to also handle the `PKGBUILD` filetype.

However, I think most configuration points for treesitter are actually configured passively by the presence of query files: highlights, indent, folding.

As an example that this could be similar too, here's what I have to configure nvim-treesitter:

```lua
require'nvim-treesitter.configs'.setup {
  ensure_installed = {
    "bash", "c", "help", "html", "json", "lua", "make"
  },
  highlight = {
    enable = true,
  },
  indent = {
    enable = true,
    is_supported = function(lang)
      return ({
        lua = true,
        c = true,
        tcl = true
      })[lang] or false
    end
  },
  incremental_selection = {
    enable = true,
    keymaps = {
      init_selection    = "gnn",
      node_incremental  = "grn",
      scope_incremental = "grc",
      node_decremental  = "grm",
    },
  },
  fold = {
    enable = true,
    disable = {'rst', 'make'}
  },
  playground = { enable = true },
  context_commentstring = { enable = true },
  query_linter = {
    enable = true,
    use_virtual_text = true,
    lint_events = {"BufWrite", "CursorMoved"},
  },
}
```

However, I'm not sure this kind of config makes sense for a language provider.

Currently, a language provider provides:
- A parser (either directly or indirectly).
- A collection of queries under `queries/{lang}/`.
- Custom directives and predicates for queries (via `plugin/` I'd imagine).
- [This PR] a directive which says what filetypes the language aims to target.

Regardless, despite whether `vim.treesitter.config()` is a good idea, I think adding something like `register` should be fine as it is more low level (and can be removed if we change our minds).

Reply:
In nvim-treesitter, we also first had the approach that the user could provide a parser config with filetype associated. Internally, we had to invert the mapping similar way as it is done in this PR, because we actually we needed a filetype to lang mapping. Users sometimes had problems configuring the parser for one specific filetype. To do that they had to remove that filetype from the one parser config and add it to the other. Often they only added it to the alternative parser config and had then semi-random behavior (depending on which parser config registered first).

We ended up exposing the formally private map and discouraged setting filetypes on parser configs.

Reply:
No, we very much don't want to copy nvim-treesitter's architecture! I think for core, it makes much more sense to be 
1. filetype-centric (i.e., you configure treesitter-based features in a ftplugin or autocommand, same as any other feature)
2. integrated into the `syntax` feature (which probably needs to be extended to cover the possibilities; e.g., by accepting a dict-like table -- once those are supported for options -- in addition to a boolean or string).

Reply:
Can't config() start out with the impl of register() here? I'm hoping we can avoid the extra surface area. This is a configuration task, after all.

Reply:
> No, we very much don't want to copy nvim-treesitter's architecture! I think for core, it makes much more sense to be

I didn't say that. I just wanted to warn that `add(lang, {filetype = { "a", "b", "c" }})` can easily generate configuration errors when a users mentions ft `c"` for multiple languages. `register` does have this problem and would preferably be called in ftplugins so a filetype can select their parser rather than having parsers select their filetype. @lewis6991 already mentioned that `register` is preferred.

The implementation of nvim-treesitter of this is horrible. I'm happy when this PR can replace it.

Reply:
> Can't config() start out with the impl of register() here? I'm hoping we can avoid the extra surface area. This is a configuration task, after all.

The existing `config` model doesn't work well with multiple sources, so I'm unsure about introducing that right now without a more thought out design. I'd assume config would exist at `vim.treesitter.config()` (higher level) as opposed to `vim.treesitter.language.register()` (low level).

As far as surface area is concerned, I think a more pragmatic approach for treesitter should be more (but not totally) white-box for the time being, I.e. more things are accessible from the outside as this allows more experimentation from downstream. We can do this until we can figure out the exact shape we want things to take, and at that point we can begin to deprecate and private parts we want to unexpose. We may even need to iterate several times in some areas (like we already have). What's important is that we maintain progress.

My desired goal is that `register` becomes internal only and that `lang` becomes a mostly unimportant detail that's only needed when managing the parsers and queries.

>  register does have this problem and would preferably be called in ftplugins so a filetype can select their parser rather than having parsers select their filetype.

You're right. I was vaguely aware of this when putting this together and considered adding extra validation to detect such conflicts. For now, I decided on a passive approach of last wins, with the prospect of tightening this up in the future if it continues to be a problem. We have a similar problem with highlight groups (mostly solved with `default`), and syntax files (kind of solved with `after`).

Reply:
> The existing `config` model doesn't work well with multiple sources, so I'm unsure about introducing that right now without a more thought out design. I'd assume config would exist at `vim.treesitter.config()` (higher level) as opposed to `vim.treesitter.language.register()` (low level).

ok agreed. 👍